### PR TITLE
Add support of ARC MWDT libc to Zephyr

### PR DESCRIPTION
**Pull request by evgeniy-paltsev**
ARC MWDT toolchain doesn't provide NEWLIB libc. Let's add support of of ARC MWDT libc to Zephyr.

This is prerequisite of adding ARC MWDT C++ library support to Zephyr.

### CODE REVIEW

**Comment by abrodkin** [comment]
Needs to be re-based on top of https://github.com/zephyrproject-rtos/zephyr/pull/36744 once it is merged.

**Review by abrodkin — APPROVED** [review]

File: lib/libc/arcmwdt/arcmwdt-string.c [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+#include <sys/types.h>
+
+/**
+ *
+ * @brief Get fixed-size string length
+ *
+ * @return number of bytes in fixed-size string <s>
+ */
+
+size_t strnlen(const char *s, size_t maxlen)
```
Comment (ruuddw) [inline lib/libc/arcmwdt/arcmwdt-string.c:20]: No strnlen in MW libc? maybe file a enhancement request for that...
I do see strnlen_s is part of MW, use that instead?

File: lib/libc/arcmwdt/include/fcntl.h [hunk 2]
```diff
@@ -0,0 +1,21 @@
+/*
+ * Copyright (c) 2021 Synopsys
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_LIB_LIBC_ARCMWDT_INCLUDE_FCNTL_H_
+#define ZEPHYR_LIB_LIBC_ARCMWDT_INCLUDE_FCNTL_H_
+
+#define O_CREAT    0x0200
```
Comment (ruuddw) [inline lib/libc/arcmwdt/include/fcntl.h:10]: Aren't these define in MW sys/fctl?
Can you simply define them here, without ensuring that the MW implementation isn't using other constants?

File: lib/libc/arcmwdt/arcmwdt-string.c (hunk 1 above)
Comment (evgeniy-paltsev) [reply lib/libc/arcmwdt/arcmwdt-string.c:13]: Fixed. Now we simply call strnlen_s in strnlen.

File: lib/libc/arcmwdt/include/fcntl.h (hunk 2 above)
Comment (evgeniy-paltsev) [reply lib/libc/arcmwdt/include/fcntl.h]: Fixed. Now we only define O_NONBLOCK which is missing in MWDT fcntl.h header. We also check that our O_NONBLOCK value doesn't conflict with any other file IO defines which came from MWDT fcntl.h header.

Comment (ruuddw) [reply lib/libc/arcmwdt/include/fcntl.h]: Nice solution, thanks.

**Review by ruuddw — APPROVED** [review]
Ready for merge imho.

**Comment by evgeniy-paltsev** [comment]
The warnings in compliance check (`WARNING:NEW_TYPEDEFS: do not add new typedefs`) are related to adding types missing in MWDT toolchain, so they should be ignored.

**Review by dcpleung — APPROVED** [review]

**Comment by evgeniy-paltsev** [comment]
This PR has enough approvals and it was kept open enough time so everyone could share their concerns (if any). I guess we can safely merge it now.

File: lib/libc/arcmwdt/libc-hooks.c [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+static int (*_stdout_hook)(int) = _stdout_hook_default;
+
+void __stdout_hook_install(int (*hook)(int))
+{
+	_stdout_hook = hook;
+}
+
+int z_impl_zephyr_fputc(int c, FILE *stream)
```
Comment (stephanosio) [inline lib/libc/arcmwdt/libc-hooks.c:31]: Here you are implicitly implementing the "minimal libc" hook interface functions:
https://github.com/zephyrproject-rtos/zephyr/blob/775cf55d9268201be6f42523c80ffad808fd3209/include/sys/libc-hooks.h#L19-L37

You should add `#elif defined(CONFIG_ARCMWDT_LIBC)` there and add only the required hook functions for the ARC MWDT libc.

**Review by stephanosio — CHANGES_REQUESTED** [review]
The `libc-hooks.c` implementation does not look sane to me.

I assume `fputs` and others are already implemented by the ARC MWDT libc, but they are declared weak and therefore overridable.

In that case, only the most fundamental (interface) functions should be overridden by the `libc-hooks.c` (e.g. `write()` and `read()` as with the newlib implementation). I highly doubt the ARC MWDT libc implementation just happens to have the same interface definition as what we currently have in `stdout_console.c` of the minimal libc.

File: lib/libc/arcmwdt/libc-hooks.c [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+#ifdef CONFIG_USERSPACE
+static inline int z_vrfy_zephyr_fputc(int c, FILE *stream)
+{
+	return z_impl_zephyr_fputc(c, stream);
+}
+#include <syscalls/zephyr_fputc_mrsh.c>
+#endif
+
+int fputc(int c, FILE *stream)
```
[the remark this answers is not in the document]
Comment (evgeniy-paltsev) [reply lib/libc/arcmwdt/libc-hooks.c]: Yep, you are correct. I've dropped it in a new implementation.

File: lib/libc/arcmwdt/libc-hooks.c (hunk 3 above)
Comment (evgeniy-paltsev) [reply lib/libc/arcmwdt/libc-hooks.c]: Fixed in new implementation, thanks!

File: include/toolchain/mwdt.h [hunk 5]
```diff
@@ -61,6 +61,17 @@
...
+#ifndef _MLIBC_RESTRICT_defined
+#define _MLIBC_RESTRICT_defined
+
+#ifdef __cplusplus
+	#define _MLIBC_RESTRICT __restrict__
+#else
+	#define _MLIBC_RESTRICT restrict
+#endif
+
+#endif /* _MLIBC_RESTRICT_defined */
```
[the remark this answers is not in the document]
Comment (evgeniy-paltsev) [reply include/toolchain/mwdt.h]: Removed.

File: lib/libc/arcmwdt/libc-hooks.c [hunk 6]
```diff
@@ -0,0 +1,113 @@
+/* stdout_console.c */
```
[the remark this answers is not in the document]
Comment (evgeniy-paltsev) [reply lib/libc/arcmwdt/libc-hooks.c]: Fixed.

**Review by stephanosio — CHANGES_REQUESTED** [review]
Looks ok overall. Some minor fixes.

File: lib/libc/arcmwdt/libc-hooks.c [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+		_stdout_hook(*(buf + i));
+	}
+	return nbytes;
+}
+
+#ifdef CONFIG_USERSPACE
+static inline int z_vrfy_z_zephyr_write_stdout(const void *buf, int nbytes)
+{
+	Z_OOPS(Z_SYSCALL_MEMORY_READ(buf, nbytes));
+	return z_impl_zephyr_write_stdout((const void *)buf, nbytes);
```
Comment (stephanosio) [inline lib/libc/arcmwdt/libc-hooks.c:48]: Cast should not be necessary.
```suggestion
	return z_impl_zephyr_write_stdout(buf, nbytes);
```

Comment (evgeniy-paltsev) [reply lib/libc/arcmwdt/libc-hooks.c]: Thanks for pointting, fixed!

File: lib/libc/arcmwdt/libc-hooks.c [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+	return z_impl_zephyr_write_stdout(buf, nbytes);
+}
+#endif
+
+/*
+ * It's require to implement _isatty to have STDIN/STDOUT/STDERR buffered
+ * properly.
+ */
+int _isatty(int file) {
+	if (file == STDIN_FILENO || file == STDOUT_FILENO || file == STDERR_FILENO) {
```
Comment (stephanosio) [inline lib/libc/arcmwdt/libc-hooks.c:67]:
```
-:308: ERROR:OPEN_BRACE: open brace '{' following function definitions go on the next line
#308: FILE: lib/libc/arcmwdt/libc-hooks.c:66:
+int _isatty(int file) {
```

```suggestion
int _isatty(int file)
{
	if (file == STDIN_FILENO || file == STDOUT_FILENO || file == STDERR_FILENO) {
```

**Review by stephanosio — APPROVED** [review]

**Review by ruuddw — APPROVED** [review]
re-approved latest version

**Comment by cfriedt** [comment]
Thanks @evgeniy-paltsev - will merge as soon as it's green. And thanks to everyone for the reviews ;-)

**Comment by evgeniy-paltsev** [comment]
@cfriedt ok, now there are only a few warnings about adding typedefs missing in MWDT toolchain, so they should be ignored.

Thanks for pointing to the issue with too long commit line!

**Pull request merged**